Add spiders

tanwayne890 opened the pull request:
This PR adds:
- `MeleeSpider`, `RangeSpider`, `IceMeleeSpider`, `IceRangeSpider`, `WaterMeleeSpider`, `WaterRangeSpider`, `FireMeleeSpider`, `FireRangeSpider` prefabs and its scriptable objects: `EnemyInfo` and `EnemyShootingInfo` respectively
- damage recalculation for elemental spiders
- show damage text when enemy takes damage

Can use existing debug log to test whether enemy receives the correct amount of damage. Note that, enemy's HP is reduced by formula: HP - (bullet damage - DEF).

May need help on a new bug:
Saw one tower bullet on enemy, but send multiple times of damage to enemy #209

tanwayne890 committed:
add elementEffectType

tanwayne890 committed:
Add elementType into enemyInfo (scriptable object), recalculate damage

tanwayne890 committed:
use example enemy: elementalMelee

tanwayne890 committed:
add all spiders and debug log

tanwayne890 committed:
add damagetext when enemy receive damage

tanwayne890 commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -71,26 +72,64 @@ public class Enemy : MonoBehaviour {
+     * 
+     * Reference: https://app.diagrams.net/#G1qh2Om2KumNYIiOz-RVrV0Wqfpf9qtsK3 
+   
+    // ElementEffectType (in ElementEffectInfo) have COMBINED element, to be used in EffectManager
+    // if possible, may need to refactor EffectManager to remove COMBINED
+    // The final outcome is enemy elements have no COMBINED element and tower elements (ElementType) have no COMBINED element.
```
FYI

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -71,26 +72,64 @@ public class Enemy : MonoBehaviour {
+            return;
+        }
+        ReduceHealth(amount);
+        Debug.Log("take damage normal: " + amount);
```
Comment out debug log

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -71,26 +72,64 @@ public class Enemy : MonoBehaviour {
+            || (elementType == ElementEffectType.WATER && bulletElementType == ElementEffectType.ICE)
+            || (elementType == ElementEffectType.ICE && bulletElementType == ElementEffectType.FIRE)) {
+            ReduceHealth(amount * 150 / 100);
+            Debug.Log("take damage more: " + amount * 150 / 100);
```
Comment out debug log

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -26,6 +26,7 @@ public class Enemy : MonoBehaviour {
     private int defense;
     private float inkGained;
     private GameObject deathEffect;
+    private ElementEffectType elementType;
```
Suggestion: Move into EnemyInfo instead of putting in the enemy class. And instead of using `ElementEffectType`, use `ElementInfo`

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -71,26 +72,64 @@ public class Enemy : MonoBehaviour {
+
+     */
+    public void TakeDamage(float amount, ElementEffectType bulletElementType) {
+        if ((elementType == ElementEffectType.FIRE && bulletElementType == ElementEffectType.FIRE)
+            || (elementType == ElementEffectType.WATER && bulletElementType == ElementEffectType.WATER)
+            || (elementType == ElementEffectType.ICE && bulletElementType == ElementEffectType.ICE)) {
+            return;
```
```suggestion
        if (elementType == bulletElementType
            && elementType != ElementEffectType.NONE && elementType != ElementEffectType.COMBINED) {
            return;

```

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -71,26 +72,64 @@ public class Enemy : MonoBehaviour {
 
-        // float number between 0 and 1
-        healthBar.fillAmount = health / enemyInfo.health;
+        if ((elementType == ElementEffectType.FIRE && bulletElementType == ElementEffectType.WATER)
+            || (elementType == ElementEffectType.WATER && bulletElementType == ElementEffectType.ICE)
+            || (elementType == ElementEffectType.ICE && bulletElementType == ElementEffectType.FIRE)) {
```
Hardcoded. Suggestion:
Add to class `ElementInfo` a `ElementType weakness`, and do a check `element.weakness.type == bulletElementType`

trxe commented on Assets/Scripts/Enemy/EnemyInfo.cs:
```diff
@@ -7,4 +7,5 @@ public class EnemyInfo : ScriptableObject {
     public int defense;
     public float inkGained;
     public GameObject deathEffect;
+    public ElementEffectType elementType;
```
change to `ElementInfo`

trxe reviewed (changes requested):
Made some suggestions to improve structure, do review and update

tanwayne890 committed:
Add `SpiderScene-prototype` (#211)

* add cemetery halloween set

* add prototype scene with spider bg, base and necessary albedo

* remove cemetery halloween set (due to high poly and colourful)

* add deadtrees, bones, skull, eggs

* add spider animator, spider models and prefabs

* Spider model fix

- Fix textures for spider models
- Fix Polygon face orientation for spider models

* gather all spider prefabs into one folder

* move materials and fbms into Assets/Materials/Enemy/Spiders

* add spider level info for testing, fix FireMeleeSpider name, adjust lighting

* remove bizulkaproduction sample scene

* remove explosiveLLC

Co-authored-by: Danh Nguyen <ducdanh24112001@gmail.com>
Co-authored-by: trxe <peixianwong@gmail.com>

tanwayne890 committed:
change damageText colour green

tanwayne890 committed:
add elementInfo into EnemyInfo, add weakness into elementInfo

tanwayne890 commented:
Fixed:
- Add to class ElementInfo a ElementType weakness
- In `EnemyInfo` & `Enemy`, ElementEffectType change to ElementInfo
- In `Enemy`, use ElementInfo to calculate received damage

trxe commented on Assets/Scripts/Enemy/Enemy.cs:
```diff
@@ -71,26 +72,58 @@ public class Enemy : MonoBehaviour {
+        }
+
+        if (element.weakness == bulletElement.type) {
+            ReduceHealth(amount * 150 / 100);
```
suggestion: make 150/100 a `[SerializableField] float multiplier`

tanwayne890 replied:
https://github.com/CS3247-Game-Development-Team-6/Doodles/pull/208/commits/143236b3b0bb0d67b3161e8271d2aad35f7ed2e5 added the multiplier into EnemyInfo (scriptable object)

tanwayne890 replied:
![image](https://user-images.githubusercontent.com/58804560/190205620-108a32e0-2327-45bd-b864-765e101c2aeb.png)

trxe reviewed (approved):
once you make the edit, lgtm. thanks

tanwayne890 committed:
add damage multiplier into EnemyInfo
